src/AppInstallerCommonCore/PortableARPEntry.cpp

```diff
@@ -0,0 +1,142 @@
+            root = HKEY_LOCAL_MACHINE;
+            if (arch == Utility::Architecture::X64)
+            {
+                subKey = s_UninstallRegistryX64;
+            }
+            else
+            {
+                subKey = s_UninstallRegistryX86;
+            }
+        }
+        else
+        {
+            // HKCU uninstall registry share the x64 registry view.
+            root = HKEY_CURRENT_USER;
+            subKey = s_UninstallRegistryX64;
+        }
+
+        subKey += L"\\" + productCode;
+        m_key = Key::OpenIfExists(root, subKey, 0, KEY_ALL_ACCESS);
+        if (m_key != NULL)
+        {
+            m_exists = true;
+        }
+        else
+        {
+            m_key = Key::Create(root, subKey);
+        }
+    }
+
+    std::wstring_view ToString(PortableValueName valueName)
+    {
+        switch (valueName)
+        {
+            VALUENAMECASE(DisplayName);
+            break;
+            VALUENAMECASE(DisplayVersion);
+            break;
+            VALUENAMECASE(Publisher);
+            break;
+            VALUENAMECASE(InstallDate);
+            break;
+            VALUENAMECASE(URLInfoAbout);
+            break;
+            VALUENAMECASE(HelpLink);
+            break;
+            VALUENAMECASE(UninstallString);
+            break;
+            VALUENAMECASE(WinGetInstallerType);
+            break;
+            VALUENAMECASE(InstallLocation);
+            break;
+            VALUENAMECASE(PortableTargetFullPath);
+            break;
+            VALUENAMECASE(PortableSymlinkFullPath);
+            break;
+            VALUENAMECASE(SHA256);
+            break;
+            VALUENAMECASE(WinGetPackageIdentifier);
+            break;
+            VALUENAMECASE(WinGetSourceIdentifier);
+            break;
+            VALUENAMECASE(InstallDirectoryCreated);
+            break;
+            default: return {};
+        }
+    }
+
+    bool PortableARPEntry::IsSamePortablePackageEntry(const std::string& packageId, const std::string& sourceId)
+    {
+        auto existingWinGetPackageId = m_key[std::wstring{ s_WinGetPackageIdentifier }];
+        auto existingWinGetSourceId = m_key[std::wstring{ s_WinGetSourceIdentifier }];
+
+        bool isSamePackageId = false;
+        bool isSamePackageSource = false;
+
+        if (existingWinGetPackageId.has_value())
+        {
+            isSamePackageId = existingWinGetPackageId.value().GetValue<Value::Type::String>() == packageId;
+        }
+
+        if (existingWinGetSourceId.has_value())
+        {
+            isSamePackageSource = existingWinGetSourceId.value().GetValue<Value::Type::String>() == sourceId;
+        }
+
+        return isSamePackageId && isSamePackageSource;
+    }
+
+    void PortableARPEntry::SetValue(PortableValueName valueName, const std::wstring& value)
+    {
+        m_key.SetValue(std::wstring{ ToString(valueName) }, value, REG_SZ);
+    }
+
+    void PortableARPEntry::SetValue(PortableValueName valueName, const std::string& value)

```

Comment:
perf: If you are going to immediately convert to UTF-16, take in a `std::string_view`.  Really "if you don't need the guaranteed nullptr at the end of the C-string".